### PR TITLE
Fix updating port notes via the API

### LINKED ISSUE #15762: API - Can't update port notes
### The problem

I can't update the port notes. I keep getting :

`Cannot access offset of type string on string {"userId":1,"exception":"[object] (TypeError(code: 0): Cannot access offset of type string on string at /opt/librenms/includes/common.php:294)"}`

curl -svk -X PATCH -H "X-Auth-Token: <TOKEN>" -H "Content-Type: application/json" -d '{"notes": "test"}' https://librenms-server/api/v0/devices/63/port/3953

I've tried different variations too. I even used the example in the notes:
https://docs.librenms.org/API/Devices/#update_device_port_notes
but it still fails. When I change the device_id to a hostname (from select hostname from devices;), I get:

```
SQLSTATE[23000]: Integrity constraint violation: 1048 Column 'device_id' cannot be null (Connection: mysql, SQL: insert into `devices_attribs` (`attrib_type`, `attrib_value`, `device_id`) values (port_id_notes:2, test, ?)) {"userId":1,"exception":"[object] (Illuminate\\Database\\QueryException(code: 23000): SQLSTATE[23000]: Integrity constraint violation: 1048 Column 'device_id' cannot be null (Connection: mysql, SQL: insert into `devices_attribs` (`attrib_type`, `attrib_value`, `device_id`) values (port_id_notes:2, test, ?)) at /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Connection.php:822)
[previous exception] [object] (PDOException(code: 23000): SQLSTATE[23000]: Integrity constraint violation: 1048 Column 'device_id' cannot be null at /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Connection.php:580)"}

```

I just want to update the port notes via the api,so it shows up ie)
/device/device=63/tab=port/port=3953/view=notes/

Any advice?

### Output of ./validate.php

```txt
./validate.php
===========================================
Component | Version
--------- | -------
LibreNMS  | 24.1.0-39-gf9fc80966 (2024-01-17T16:58:08-05:00)
DB Schema | 2023_12_12_171400_alert_rule_note (277)
PHP       | 8.1.2-1ubuntu2.14
Python    | 3.10.12
Database  | MariaDB 10.6.12-MariaDB-0ubuntu0.22.04.1
RRDTool   | 1.7.2
SNMP      | 5.9.1
===========================================

[OK]    Composer Version: 2.6.6
[OK]    Dependencies up-to-date.
[OK]    Database connection successful
[OK]    Database Schema is current
[OK]    SQL Server meets minimum requirements
[OK]    lower_case_table_names is enabled
[OK]    MySQL engine is optimal
[OK]    Database and column collations are correct
[OK]    Database schema correct
[OK]    MySQL and PHP time match
[OK]    Active pollers found
[OK]    Dispatcher Service not detected
[OK]    Locks are functional
[OK]    Python poller wrapper is polling
[OK]    Redis is unavailable
[OK]    rrd_dir is writable
[OK]    rrdtool version ok
```

### What was the last working version of LibreNMS?

_No response_

### Anything in the logs that might be useful for us?

_No response_

### PR DESCRIPTION
**Pull request by laf**
Fixes #15762

This fix was recommended in a rejected PR but never submitted so just getting it done.

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [ ] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [ ] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [ ] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Review by PipoCanaja — APPROVED** [review]
LGTM but not tested

**Comment by laf** [comment]
Tested:

```
curl -X PATCH -H "X-Auth-Token: TOKEN" -H "Content-Type: application/json" -d '{"notes": "test"}' https://demo.librenms.org/api/v0/devices/1/port/2 
{
    "status": "ok",
    "message": "Port notes field has been updated"
}
```

File: includes/html/api_functions.inc.php [hunk 1]
```diff
@@ -1262,6 +1262,7 @@ function update_device_port_notes(Illuminate\Http\Request $request): JsonRespons
     $hostname = $request->route('hostname');
     // use hostname as device_id if it's all digits
     $device_id = ctype_digit($hostname) ? $hostname : getidbyname($hostname);
+    $device = device_by_id_cache($device_id);
```
Comment (murrant) [inline includes/html/api_functions.inc.php:1265]:
```suggestion
    $device = DeviceCache::get($hostname);
```

File: includes/html/api_functions.inc.php [hunk 2]
```diff
@@ -1270,7 +1271,7 @@ function update_device_port_notes(Illuminate\Http\Request $request): JsonRespons
         return api_error(400, 'Port field to patch has not been supplied.');
     }
 
-    if (set_dev_attrib($device_id, 'port_id_notes:' . $portid, $content)) {
+    if (set_dev_attrib($device, 'port_id_notes:' . $portid, $content)) {
```
Comment (murrant) [inline includes/html/api_functions.inc.php:1274]:
```suggestion
    if ($device->setAttrib('port_id_notes:' . $portid, $content)) {
```

**Comment by laf** [comment]
Thanks for the suggestions. Tested after the updates and this still works

```
curl -X PATCH -H "X-Auth-Token: TOKEN" -H "Content-Type: application/json" -d '{"notes": "test"}' https://demo.librenms.org/api/v0/devices/1/port/2 
{
    "status": "ok",
    "message": "Port notes field has been updated"
}
```

**Pull request merged**